app/models/service/resource_linking.rb

```diff
@@ -0,0 +1,37 @@
+module Service::ResourceLinking
+  extend ActiveSupport::Concern
+
+  def add_provider_vms(options)
+    # required keys in options hash:
+    #   :target_id, the id of the provider to refresh
+    #   :uid_ems_array, an array of VMs' uid_ems
+    #
+    options = {
+      :target_class => "ExtManagementSystem",
+      :name         => "Linking VMs to service #{name} ID: [#{id}]",
+      :userid       => evm_owner.userid,
+      :sync_key     => guid,
+      :service_id   => id,
+      :zone         => my_zone
+    }.merge(normalize_inputs!(options))
+
+    _log.info("NAME [#{options[:name]}] for user #{evm_owner.userid}")
+
+    Service::LinkingWorkflow.create_job(options)
+  rescue => err
+    _log.log_backtrace(err)
+    raise
+  end
+
+  private
+
+  def normalize_inputs!(options)
+    raise _("no target_id defined for linking to service") unless options[:target_id]
+
+    uid_ems_array = options[:uid_ems_array] || []
+    options[:uid_ems_array] = uid_ems_array.kind_of?(Array) ? uid_ems_array.compact.uniq : [uid_ems_array]
```

Comment:
Not sure if the suggestion to reduce this to these two lines to the follow was lost in the comments or if you are preferring the current approach.

Suggested change: `options[: uid_ems_array] = Array(options[: uid_ems_array]).compact.uniq`